beebop/utils.py

```diff
@@ -141,45 +138,81 @@ def get_external_clusters_from_file(
     :param external_clusters_prefix: prefix for external cluster name
-    :return dict: [dict of sample hash to lowest numbered-cluster for that
-        sample]
-    """
-    # copy hashes list to keep track of hashes we still need to find samples
-    # for
-    remaining_hashes = hashes_list[:]
-    result = {}
-    with open(previous_query_clustering_file) as f:
-        reader = csv.reader(f, delimiter=",")
-        for row in reader:
-            # We expect two columns in the external clusters csv: the
-            # first contains the sample id (which will be the hash in
-            # the case of our uploaded samples), and the second contains
-            # all the external cluster numbers for the sample, separated
-            # by semicolons
-            sample_id = row[0]
-            if sample_id in remaining_hashes:
-                # Add lowest numeric cluster to dictionary
-                if len(row) > 1:
-                    cluster_no = get_lowest_cluster(row[1])
-                    result[sample_id] = (
-                        f"{external_clusters_prefix}{cluster_no}"
-                    )
-
-                # Remove sample id from remaining hashes to find
-                remaining_hashes.remove(sample_id)
-
-                # Break if no hashes left to find
-                if len(remaining_hashes) == 0:
-                    break
-    return result
+    :return tuple: [dictionary of sample hash to external cluster name,
+        list of sample hashes that were not found]
+    """
+    df, samples_mask = get_df_sample_mask(
+        previous_query_clustering_file, hashes_list
+    )
+    filtered_df = df[samples_mask]
+
+    # Split into found and not found based on NaN values
+    found_mask = filtered_df["Cluster"].notna()
+    not_found_hashes = filtered_df[~found_mask]["sample"].tolist()
+
+    # Process only rows with valid clusters
+    valid_clusters = filtered_df[found_mask]
+    cluster_numbers = valid_clusters["Cluster"].apply(get_lowest_cluster)
+
+    hash_to_cluster_mapping = (
+        external_clusters_prefix + cluster_numbers.astype(str)
+    )
+    hash_to_cluster_mapping.index = valid_clusters["sample"]
+
+    return hash_to_cluster_mapping.to_dict(), not_found_hashes
+
+
+def update_external_clusters_csv(
+    previous_query_clustering_file: str,
+    not_found_q_names: list,
+    external_clusters_not_found: dict,
+) -> None:
+    """
+    [Update the external clusters CSV file with the clusters of the samples
+    that were not found in the external clusters file.]
+
+    :param previous_query_clustering_file: [Path to CSV file
+    containing sample data]
+    :param not_found_q_names: [List of sample names
+    that were not
+    found in the external clusters file]
+    :param external_clusters_not_found: [Dictionary mapping
+    sample names to external cluster names]
+    """
+    df, samples_mask = get_df_sample_mask(
+        previous_query_clustering_file, not_found_q_names
+    )
```

Comment:
So this is going to get a dataframe of the whole file, plus a mask or view of the rows which match the not_found samples so you can easily update those?